src/wallet/wallet2.cpp

```diff
@@ -4597,18 +4700,117 @@ void wallet2::setup_keys(const epee::wipeable_string &password)
     m_account.decrypt_viewkey(key);
   }
 
-  m_cache_key = derive_cache_key(key);
+  m_cache_key = derive_cache_key(key, config::HASH_KEY_WALLET_CACHE);
 
   get_ringdb_key();
 }
 //----------------------------------------------------------------------------------------------------
+void validate_background_cache_password_usage(const tools::wallet2::BackgroundSyncType background_sync_type, const boost::optional<epee::wipeable_string> &background_cache_password, const bool multisig, const bool watch_only, const bool key_on_device)
+{
+  THROW_WALLET_EXCEPTION_IF(multisig || watch_only || key_on_device, error::wallet_internal_error, multisig
+      ? "Background sync not implemented for multisig wallets" : watch_only
+      ? "Background sync not implemented for view only wallets"
+      : "Background sync not implemented for HW wallets");
+
+  switch(background_sync_type)
+  {
+    case tools::wallet2::BackgroundSyncOff:
+    {
+      THROW_WALLET_EXCEPTION(error::wallet_internal_error, "background sync is not enabled");
+      break;
+    }
+    case tools::wallet2::BackgroundSyncReusePassword:
+    {
+      THROW_WALLET_EXCEPTION_IF(background_cache_password, error::wallet_internal_error,
+          "unexpected custom background cache password");
+      break;
+    }
+    case tools::wallet2::BackgroundSyncCustomPassword:
+    {
+      THROW_WALLET_EXCEPTION_IF(!background_cache_password, error::wallet_internal_error,
+          "expected custom background cache password");
+      break;
+    }
+    default: THROW_WALLET_EXCEPTION(error::wallet_internal_error, "unknown background sync type"); break;
+  }
+}
+//----------------------------------------------------------------------------------------------------
+typedef boost::variant<epee::wipeable_string, crypto::chacha_key> background_key_derivation_t;
+void get_background_key(const background_key_derivation_t &background_key_derivation, crypto::chacha_key &background_key, const uint64_t kdf_rounds)
+{
+  crypto::chacha_key key;
+  if (background_key_derivation.type() == typeid(epee::wipeable_string))
+  {
+    const epee::wipeable_string password = boost::get<epee::wipeable_string>(background_key_derivation);
+    crypto::generate_chacha_key(password.data(), password.size(), key, kdf_rounds);
+
+    // Re-deriving the wallet cache key supports being able to use either a
+    // background cache password OR a normal password as the wallet's password
+    // when opening a background wallet
+    key = derive_cache_key(key, config::HASH_KEY_WALLET_CACHE);
```

Comment:
It's thanks to this line that a user who chooses Option 1 (`reuse-wallet-password`) is still able to open their background wallet file directly, using their wallet password, which thus exposes the password which can be used to decrypt their encrypted spend key (which lives on disk).

@DangerousFreedom1984 pointed out this could potentially lead a user to leave their spend key more exposed than the user may realize. I think this is a fair assessment.

I also have concern over a user who chooses Option 2 (`custom-background-password`) and ends up providing a custom password that is the same as their wallet password, which has the same effect as the above.

Couple more TODO's:
1. When a user chooses Option 1 (`reuse-wallet-password`), remove the ability for a user to open a background wallet directly reusing their wallet password. Thus, opening a background wallet directly will only be supported via Option 2.
2. When a user chooses Option 2 (`custom-background-password`), ensure the provided custom password is not the same as the user's wallet password.